doc/POINTS_COORDINATES.md

```diff
@@ -0,0 +1,169 @@
+relative so some larger scale.  For example, when performing mapgen for a
+single overmap, we want to work with coordinates within that overmap.  This
+will be an overmap terrain-scale point relative to the corner of its containing
+overmap, and so typically take `x` and `y` values in the range [0,180).
+
+## Vertical coordinates
+
+Although `x` and `y` coordinates work at all these various scales, `z`
+coordinates are consistent across all contexts.  They lie in the range
+[-`OVERMAP_DEPTH`,`OVERMAP_HEIGHT`].
+
+## Vehicle coordinates
+
+TODO.
+
+## Point types
+
+To work with these coordinate systems we have a variety of types.  These are
+defined in [`coordinates.h`](../src/coordinates.h).  For example, we have
+`point_abs_ms` for absolute map-square coordinates.  The three parts of the
+type name are *dimension*`_`*origin*`_`*scale*.
+
+* **dimension** is either `point` for two-dimensional or `tripoint` for
+  three-dimensional.
+* **origin** specifies what the value is relative to, and can be:
+  * `rel` means relative to some arbitrary point.  This is the result of
+    subtracting two points with a common origin.  It would be used for example
+    to represent the offset between the avatar and a monster they are shooting
+    at.
+  * `abs` means global absolute coordinates.
+  * `sm` means relative to a corner of a submap.
+  * `omt` means relative to a corner of an overmap terrain.
+  * `om` means relative to a corner of an overmap.
+* **scale** means the scale as discussed above.
+  * `ms` for map square.
+  * `sm` for submap.
+  * `omt` for overmap terrain.
+  * `seg` for segment.
+  * `om` for overmap.
+
+## Raw point types
+
+As well as these types with origin and scale encoded into the type, there are
+simple raw point types called just `point` and `tripoint`.  These can be used
+when no particular game scale is intended.
+
+At time of writing we are still in the process of transitioning the codebase
+away from using these raw point types everywhere, so you are likely to see
+legacy code using them in places where the more type-safe points might seem
+appropriate.
+
+## Converting between point types
+
+### Changing scale
+
+To change the scale of a point without changing its origin, use `project_to`.
+For example:
+
+```c++
+point_abs_ms pos_ms = get_avatar()->global_square_location().xy();
+point_abs_omt pos_omt = project_to<coords::omt>( pos_ms );
+assert( pos_omt == get_avatar()->global_omt_location().xy() );
+```
+
+The same function `project_to` can be used for scaling up or down.  When
+converting to a coarser coordinate system precision is of course lost.  If you
+care about the remainder then you can instead use `project_remain`.
+
+`project_remain` allows you to convert to a coarser coordinate system and also
+capture the remainder relative to that coarser point.  It returns a helper
+struct intended to be used with `std::tie` to capture the two parts of the
```

Comment:
is this like a std::pair only you can use named variables instead of .first and .second?

Reply:
Sort of, yes.

> "ms" and "sm" are dangerously similar.
> 
> How about "local map" "lm" or "tile map" "tm".

Definitely don't want to use "local" here; that already means something else.  It might make sense to just use "tile" `tl`.  But that's a pretty massive change that's definitely out of scope for this documentation PR.